[RFC] Update the Authorization specification for MCP servers (#284)
modelcontextprotocol/modelcontextprotocol: docs/specification/2025-03-26/basic/authorization.mdx

As a side-effect, this also simplifies consent flows, and services that have multiple MCP Servers
won't require individual registration per client:server pair.

## 2. Authorization Flow

### 2.1 Overview

1. Authorization servers **MUST** implement OAuth 2.1 with appropriate security measures
   (see [RFC 9700](https://datatracker.ietf.org/doc/html/rfc9700)) for both confidential and
   public clients.

2. Authorization servers **SHOULD** support the OAuth 2.1 Dynamic Client Registration Protocol
   ([RFC 7591](https://datatracker.ietf.org/doc/html/rfc7591)).

3. Authorization servers **SHOULD** implement and MCP clients **MUST** consume OAuth 2.0 Authorization
   Server Metadata ([RFC 8414](https://datatracker.ietf.org/doc/html/rfc8414)).

### 2.2 Client OAuth 2.1 Authorization

When authorization is required and not yet proven by the client, servers **MUST** respond with
*HTTP 401 Unauthorized*.

A *HTTP 401 Unauthorized* response **MUST** include a `WWW-Authenticate` header with
the Authorization Server's issuer URL that the MCP Server is using in the following format:

`WWW-Authenticate: Bearer realm="default", issuer="https://authserver.example",`  

> `scope="mcp_access_tool1", resource="https://mcp-server.example"`  

Reviewer: What is the purpose of including the `resource` parameter here? Isn't this always going to be the same hostname as the server this request was made to?

Author: Not relevant anymore with move to PRM.

`WWW-Authenticate: DPoP realm="default", issuer="https://authserver.example",`  
`scope="mcp_access_tool1", resource="https://mcp-server.example",`  
`algs="ES256 PS256"`

An MCP client **SHOULD** evaluate or permit the user to evaluate the trustworthiness of issuers
discovered this way (to prevent phishing). For example, an MCP client could prompt the user "Do you
want to login with `https://authserver.example`?"

An MCP client **MUST** use the `issuer` identifier to construct the OAuth 2.0 metadata discovery
document path, following
[RFC 8414](https://datatracker.ietf.org/doc/html/rfc8414#section-3).

An MCP client **MAY** use the `registration_endpoint` to register itself with OAuth Provider. An MCP
client **MAY** also be pre-registered with an OAuth Provider.

> An MCP client **MUST** check the `resource` identifier against the hostname of the MCP server. An MCP

Author: Updated this - no longer relevant with move to PRM.

client **MUST** send the `resource` identifier to the OAuth Provider, following
[RFC 8707](https://www.rfc-editor.org/rfc/rfc8707.html).

An MCP client **SHOULD** implement Demonstrating Proof of Possession (DPoP) and use it to the extent
supported by the OAuth Provider. See
[RFC 9449](https://datatracker.ietf.org/doc/html/rfc9449).

Clients initiate an
[OAuth 2.1 (draft)](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12) authorization
flow after receiving a *HTTP 401 Unauthorized* response.

### 2.3 MCP Server Metadata Discovery

MCP servers **MUST** return the fully-qualified domain name of the issuer in `WWW-Authenticate`
headers in an error respose.

MCP servers **MUST** follow [RFC 6750 Section 3.1](https://www.rfc-editor.org/rfc/rfc6750#section-3.1) for
any authorization failures.

For MCP server capability discovery:

- MCP clients **MUST** parse the `WWW-Authenticate` response to discover the issuer URL, and construct
  the metadata endpoint from that issuer identifier.  

> - MCP servers **SHOULD** use a third-party authorization server and include its issuer in

Reviewer: It seems weird to mention third-party here. I like the mention of using an external authorization server in the intro, but it seems wrong to include it in the normative language in the discovery step here since it's just an implementation detail of the OAuth deployment.

Reviewer 2: +1 I have been reading this as "First Party" or "External". That also accounts for scenarios where the AS and the RS might be co-located. 

https://github.com/modelcontextprotocol/modelcontextprotocol/pull/284#issuecomment-2785092195

Author: Updated to clarify the intent.

  providers.  
- Authorization Servers **MUST** support
  [OAuth 2.0 Authorization Server Metadata](https://datatracker.ietf.org/doc/html/rfc8414)
- MCP servers **MUST** respond to an unauthorized request with a `WWW-Authenticate` header containing
  `issuer`, `scope`, and `resource` identifiers.

The discovery flow is illustrated below:  

```mermaid
sequenceDiagram
    participant MC as MCP Client
    participant MS as MCP Server
    participant AUTHZS as Authorization Server

    MC->>MS: Request data
    MS-->>MC: 401 Unauthorized with WWW-Authenticate
    MC->>MC: Construct metadata endpoint from issuer identifier

    Note over MC: Continue with authorization flow
```

#### 2.3.1 Server Metadata Discovery Headers

MCP clients **SHOULD** include the header `MCP-Protocol-Version: <protocol-version>` during Server
Metadata Discovery to allow the MCP server to respond based on the MCP protocol version.

For example: `MCP-Protocol-Version: 2024-11-05`

### 2.4 Authorization server integration

#### 2.4.1 Overview

MCP servers **MAY** override endpoints on an as-needed basis if the authorization server requires
additional gestures that are not captured by the specification.

> MCP servers **MUST NOT** complete PKCE or other public client flows on behalf of the client.

Reviewer: PKCE Is not a public client flow, it is also used by confidential clients.

What is the purpose of this sentence? It seems like you've seen some behavior you're trying to prohibit, but this doesn't really make any sense.

Author: Updated and clarified the wording.

MCP servers that **MUST** implement the [metadata discovery process](#23-server-metadata-discovery)

> if they do not use a third-party authorization server.

Reviewer: I would remove this exception and require all clients use the discovery process. This is not about first-party or third-party authorization servers, it's just about the authorization server at all. If the MCP server wants to advertise the authorization endpoint as `https://mcp-server.example/authorize` it should still be doing that in the OAuth server metadata file.

#### 2.4.2 Flow Description

The third-party authorization flow comprises these steps:

1. MCP client requests data from the MCP server.  
1. MCP server responds with *HTTP 401 Unauthorized* and `WWW-Authenticate` header containing the
   `issuer`, `scope`, and `resource`.  

> 1. MCP client retrieves the `/authorize` endpoint from the `issuer` metadata document.  

Reviewer: ```suggestion
1. MCP client retrieves the `authorization_endpoint` from the `issuer` metadata document.  
```

Having the slash in this makes it look like the path is hardcoded. The property in the metadata file is `authorization_endpoint`

> 1. MCP client redirects user to authorization server's `/authorize` endpoint. MCP client uses `scope` and

Reviewer: ```suggestion
1. MCP client redirects user to authorization server's authorization endpoint. MCP client uses `scope` and
```

>    `resource` parameters provided in the `WWW-Authenticate` header in the `/authorize` URL.  

Reviewer: ```suggestion
   `resource` parameters provided in the `WWW-Authenticate` header in the authorization URL.  
```

1. User authorizes with authorization server with `scope` and `resource`.  
1. Authorization server redirects back to MCP client with authorization code.  
1. MCP client exchanges the code for the requested tokens.  
1. MCP client caches the token.  
1. MCP clients can now request data from the MCP server with the token attached in the `Authorization`
   header.

```mermaid
sequenceDiagram
    participant UA as User-Agent (Browser)
    participant MC as MCP Client
    participant MS as MCP Server
    participant AUTHZS as Authorization Server

    MC->>MS: Request data
    MS-->>MC: 401 Unauthorized with WWW-Authenticate

>     Note over MC: MCP Client retrieves /authorize endpoint URI from the issuer metadata document (Section 2.3)

Reviewer: ```suggestion
    Note over MC: MCP Client retrieves authorization endpoint from the issuer metadata document (Section 2.3)
```

>     MC-->>UA: Initiate authorization request at /authorize

Reviewer: ```suggestion
    MC-->>UA: Initiate authorization request at the authorization endpoint
```

    Note over UA: User authenticates and authorizes client for scope and resource

>     UA->>MC: Return auth code to client

Reviewer: ```suggestion
    UA->>MC: Return authorization code to client
```

    MC->>AUTHZS: Exchange code for token

>     AUTHZS-->>MC: Third-party access token

Reviewer: ```suggestion
    AUTHZS-->>MC: Access token
```

This is not a third-party token, that just confuses things.

### 1.1 Purpose and Scope

The Model Context Protocol provides authorization capabilities at the transport level, enabling MCP
clients to make requests to restricted MCP servers on behalf of resource owners. This specification
defines the authorization flow for HTTP-based transports.

The scope of this document is **user authorization** - how a client can authenticate a user to access a
server and the server determine whether to authorize the user to access its tools, resources, or any
other capabilities.

This document **does not discuss server-to-server authorization patterns**.

### 1.2 Baseline tenets

> 1. **MCP server developers are not security experts**. The current specification assumes that developers

Reviewer: it seems weird to mention the current spec this way in this PR? Shouldn't the PR remove this language because this PR is supposed to solve this problem?

Author: Good point, I will adjust.

   will be building their own authorization server. That is not something they ever need to do in the
   context of MCP servers. This also introduces room for serious mistakes that can lead to compromise.
1. **MCP clients are Authorization Server (AS)-agnostic**. We do not and will not expect clients to be
   "enlightened" in any capacity to support a protected MCP server. This will fragment the ecosystem.
   The specification explicitly assumes that clients are not enabled in any unique way to support
   specific Authorization Servers.
    - **Note:** It is acceptable for clients to be enhanced for specific Authorization Servers,
      primarily through pre-registration.
1. **User Authentication is not a MCP Server concern.** The MCP server can guide the client

>    to the AS but will not be responsible for authenticating the user and obtaining credential

Reviewer: It is ambiguous whether "will not be responsible for" is referring to the client or the server. 

```suggestion
   to the AS but the MCP Server will not be responsible for authenticating the user and obtaining credential
```

>    artifacts for the user. The MCP server might transform incoming credential artifacts, like tokens, but

Reviewer: This mention of transforming credential artifacts is confusing. I would just remove it. There isn't any transformation of tokens involved in this really.

and instead relies on developers adopting dedicated and well-tested third-party AS implementations.

This specification also enables existing deployments to leverage their current infrastructure.
As a side-effect, this also simplifies consent flows, and services that have multiple MCP Servers
won't require individual registration per client:server pair.

## 2. Authorization Flow

### 2.1 Overview

1. Authorization servers **MUST** implement OAuth 2.1 with appropriate security measures
   (see [RFC 9700](https://datatracker.ietf.org/doc/html/rfc9700)) for both confidential and
   public clients.

2. Authorization servers **SHOULD** support the OAuth 2.1 Dynamic Client Registration Protocol
   ([RFC 7591](https://datatracker.ietf.org/doc/html/rfc7591)).

3. Authorization servers **SHOULD** implement and MCP clients **MUST** consume OAuth 2.0 Authorization
   Server Metadata ([RFC 8414](https://datatracker.ietf.org/doc/html/rfc8414)).

### 2.2 Client OAuth 2.1 Authorization

When authorization is required and not yet proven by the client, servers **MUST** respond with
*HTTP 401 Unauthorized*.

> A *HTTP 401 Unauthorized* response **MUST** include a `WWW-Authenticate` header with information about

Reviewer: ```suggestion
A *HTTP 401 Unauthorized* response **MUST** include a `WWW-Authenticate` header with
```

> the issuer that the server is using to authenticate in the following format:

Reviewer: This is the first mention of the term "issuer", so should be clarified.

```suggestion
the Authorization Server's issuer URL that the MCP Server is using in the following format:
```

As a side-effect, this also simplifies consent flows, and services that have multiple MCP Servers
won't require individual registration per client:server pair.

## 2. Authorization Flow

### 2.1 Overview

1. Authorization servers **MUST** implement OAuth 2.1 with appropriate security measures
   (see [RFC 9700](https://datatracker.ietf.org/doc/html/rfc9700)) for both confidential and
   public clients.

2. Authorization servers **SHOULD** support the OAuth 2.1 Dynamic Client Registration Protocol
   ([RFC 7591](https://datatracker.ietf.org/doc/html/rfc7591)).

3. Authorization servers **SHOULD** implement and MCP clients **MUST** consume OAuth 2.0 Authorization
   Server Metadata ([RFC 8414](https://datatracker.ietf.org/doc/html/rfc8414)).

### 2.2 Client OAuth 2.1 Authorization

When authorization is required and not yet proven by the client, servers **MUST** respond with
*HTTP 401 Unauthorized*.

A *HTTP 401 Unauthorized* response **MUST** include a `WWW-Authenticate` header with
the Authorization Server's issuer URL that the MCP Server is using in the following format:

```http
WWW-Authenticate: Bearer realm="default", issuer="https://authserver.example",

> scope="mcp_access_tool1", resource="https://mcp-server.example"

Reviewer 3: nit: would be nice to show `scope` as a space-delimited list in the example to demonstrate how multiple scopes could be requested 
```suggestion
scope="mcp:read mcp:write", resource="https://mcp-server.example"
```

```

If the client supports DPoP and the authorization servers wants to require DPoP, the
response **MUST** follow this format:

```http
WWW-Authenticate: DPoP realm="default", issuer="https://authserver.example",

> scope="mcp_access_tool1", resource="https://mcp-server.example",

Reviewer 3: ```suggestion
scope="mcp:read mcp:write", resource="https://mcp-server.example",
```

    participant AS as Authorization Server

    MC->>MS: Request data
    MS-->>MC: 401 Unauthorized with WWW-Authenticate
    MC->>MS: Fetch PRM document
    MS-->>MC: PRM document response
    Note over MC: Validate PRM document<br/>metadata
    Note over MC: Build AS metadata URL
    MC->>AS: Fetch AS metadata
    AS-->>MC: AS metadata response

    Note over MC,AS: Continue with authorization flow
```

#### 2.3.1 Server Metadata Discovery Headers

MCP clients **SHOULD** include the header `MCP-Protocol-Version: <protocol-version>` during Server
Metadata Discovery to allow the MCP server to respond based on the MCP protocol version.

For example:

```http
MCP-Protocol-Version: 2024-11-05
```

### 2.4 Authorization server integration

#### 2.4.1 Overview

MCP clients **SHOULD** be considered _public clients_ and MCP servers **MUST** be
considered _confidential clients_, as defined in
[RFC 6749 Section 2.1](https://datatracker.ietf.org/doc/html/rfc6749#section-2.1).

MCP servers **MUST NOT** complete public client flows on behalf of the MCP client (e.g., take 
authorization codes and exchange them for tokens with a public client ID, and then pass the tokens 
back to the client).

> Public client flows **MUST** be completed by MCP Client using OAuth 2.1 Authorization

Reviewer 4: When the MCP client is a limited input device, or a background service that has no direct user interface, its not possible to authorize the user with authz code grant. In such cases, supporting other grant types such as CIBA and Device Code can be helpful. Therefore, shall we consider expanding the list of allowed grant types to CIBA and device code as well?

We recently opened a discussion on this: https://github.com/modelcontextprotocol/modelcontextprotocol/discussions/298

Reviewer 4: Or, without explicitly indicating the grant types, the spec can be flexible on the grant types to be used.

```suggestion
Public client flows **SHOULD** be completed by MCP Client using OAuth 2.1 Authorization
```

Author: I am _generally_ not a fan of non-input device flows because they are easily phishable. @pavinduLakshan - can you elaborate on a concrete end-to-end user scenario where the MCP client won't be able to request user input through the browser where alternative grants are needed?

I feel like there some thought needs to be given to _MCP server-to-MCP server_ scenarios, where there is no input available, but that's a slightly different problem that's outside the scope of _this_ RFC.

Reviewer 5: @localden I’ve been developing an AI agent that runs in the background to monitor ticket availability. When a user-specified type of ticket becomes available, the agent needs to purchase it. However, we must obtain the user’s consent before proceeding with the purchase.

To handle this, we use Client-Initiated Backchannel Authentication (CIBA). I believe this represents a legitimate use case. We shouldn't limit ourselves to using only the Authorization Code flow—we should also allow clients to implement other grant types where appropriate. @aaronpk wdyt?

Reviewer 6: This opens the possibility for flows like the draft [Pre-authorized Code Flow(https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0-ID1.html#section-3.5) as well.

Author: @max-stytch client credentials grant is distinctly different, though. It can be helpful for server-to-server communication (no human in the loop). This is about client-to-server (althought the distinction might seem a bit blurry), where there is a human in the loop, but no browser.

"No-input" flows have been generally significantly less secure, and enterprises do a lot of things to block them outright due to phishing risks ([see example comms on this](https://techcommunity.microsoft.com/blog/microsoftteamsblog/policy-changes-for-microsoft-teams-devices-using-device-code-flow-authentication/4399337), including [risks from nation-state actors](https://www.microsoft.com/en-us/security/blog/2025/02/13/storm-2372-conducts-device-code-phishing-campaign/)).

I would vote no on this personally, but input from others is appreciated.

Reviewer 3: @localden the [current specification](https://modelcontextprotocol.io/specification/2025-03-26/basic/authorization#2-1-1-oauth-grant-types:~:text=MCP%20servers%20SHOULD%20support%20the%20OAuth%20grant%20types%20that%20best%20align%20with%20the%20intended%20audience.) states 

> MCP servers SHOULD support the OAuth grant types that best align with the intended audience

Do you want the spec to limit this to only the `auth_code` and `refresh_token` grant types?

> "No-input" flows have been generally significantly less secure, and enterprises do a lot of things to block them

I believe the grant types available should be determined by the AS and the AS's threat model - a Microsoft AS is going to have dramatically different concerns from a Spotify AS, which might want to expose a CIBA or Device Auth flow in addition to auth code. It is up to the MCP Client to determine which of these flows to undertake.

Reviewer 7: This specification should just be defining minimum interoperability requirements for an Authorization Server (e.g MUST support public client with PKCE authorization grant).  The AS is free to implement additional grants and features.  The separation of roles between AS and RS means the RS shouldn't be coupled to grant types used to obtain access tokens.   Adding support for a confidential client profile makes sense but it should have its own distinct requirements from public client (e.g two different profiles).

- Profile for MCP Client as Public Client (includes min requirements for an AS)
- Profile for MCP Server as Resource Server (includes min requirements for an AS)

The use of OAuth 2.0 Device Authorization Grant should be left to AS implementation/deployment IMHO.  Its not required by this interoperability spec so MCP Clients can't rely on on its use but we shouldn't prevent an AS from supporting and a custom built MCP client using.   

The use of CIBA by a MCP Server would be out-of-scope of this specification.  The MCP Server is the client in a CIBA flow and communication is out-of-band of MCP Client to MCP Server interaction which this specification defines

Author: @max-stytch I don't disagree with the fact that different AS implementations have different concerns. Where I am looking at is _client implementation burden_. Server can choose whatever grants it needs. the question then becomes - how does that reflect on the client? Do we expect clients to implement every OAuth grant type support (e.g., polling for DCF)?

I think the spec still leaves enough room for clients to do what they need, but they **MUST** support PKCE. Maybe that's the wording that we need to change here, something along the lines of:

>MCP clients **MUST** support authorization code flow with PKCE, per OAuth 2.1 specification.
>
>MCP clients **MAY** choose to implement other authorization grants.

Reviewer 8: @shashimalcse In your example of a ticketing agent that needs to be an impromptu decision about buying something, something like CIBA for getting the user's consent makes perfect sense. But, I think that is solidly outside the scope of this spec -- this spec is about the _MCP client_ (your agent) authorizing with the _MCP server_, which would probably happen at onboarding or setup time. What happens at runtime (CIBA) feels like an implementation detail of a particular tool, not how the client authorizes with the server.

Reviewer 9: I'd rather see this spec go through without specifying every possible combination, but requiring a min flow, primarily agreeing with @nbarbettini , @mcguinness, and @localden. My read also tells me there is enough room in the spec for other flows be implemented, should an implemented wishes to do so.

Author: I updated the specification with wording that gives room for client to implement other grant types if they desire, but keep PKCE as a baseline.

AS (to the MCP client).

#### 2.5.1 Special considerations

MCP servers that use delegated authorization or implement a colocated AS **MUST**:

1. Maintain secure mapping between third-party tokens and issued client sessions.
1. Never pass third-party tokens directly to the client.
1. Validate token status before granting any data access.
1. Implement appropriate token lifecycle management.  
1. Provide appropriate responses to the client if re-authentication is needed. Please refer to
   [RFC 9470 OAuth 2.0 Step Up Authentication Challenge Protocol](https://www.rfc-editor.org/rfc/rfc9470.html).

### 2.6 Implementing authorization server overrides

In certain scenarios, AS might have limited capability support, such as no dynamic client registration, and
MCP server developers may choose to implement those capabilities themselves.

In this scenario, the server **MUST** return its fully-qualified domain name as the AS in the PRM document,
that will be linked from the `WWW-Authenticate` header in the `HTTP 401 Unauthorized` response.

The MCP server will need to be configured in accordance with

> [Section 2.4.3](#243-implementing-custom-authorization-servers) to act as an AS.

Reviewer 10: Section 2.4.3 is missing

    alt External authorization server
        MC->>AS: GET /.well-known/oauth-authorization-server
        AS-->>MC: 200 OK + Metadata Document
        Note over MC: Use endpoints from metadata
    else Co-located authorization server
        MC->>MS: GET /.well-known/oauth-authorization-server
        MS-->>MC: 200 OK + Metadata Document
        Note over MC: Use endpoints from metadata
    end

    alt Client is pre-registered with the authorization server
    else Authorization server supports dynamic client registration
        MC->>AS: Request client registration
        AS-->>MC: Confirm client registration
    else Authorization server does not support dynamic client registration
        MC->>MS: Request client registration
        MS->>AS: Register client
        AS-->>MS: Return client registration
        MS-->>MC: Return client registration
    end

    Note over MC: Generate PKCE artifacts
    MC-->>UA: Initiate authorization call
    Note over UA: User authenticates and authorizes<br/>client for scope and resource
    UA->>MC: Return auth code to client
    MC->>AS: Exchange auth code for tokens
    AS-->>MC: Return tokens
    Note over MC: Store tokens

>     Note over MC, MS: Begin standard MCP message exchange

Reviewer 11: Should we mention explicitly that the messages include the token? This is already noted in 2.4.2, and it's reasonable to assume it's understood here, but perhaps we should duplicate that here.

Related: "begin" might imply something about the initial SSE connection and not subsequent requests. I mention it because it seems some clients only use configured headers/QSPs on the initial connection, but drop them for subsequent messages. That's probably a bug on their end, but it still makes me think additional prescriptivism around using the token for all messages might be beneficial.

Something similar might be good to have in 2.9 as well.

  servers they are connecting to.
- Authorization servers **MUST** support
  [OAuth 2.0 Authorization Server Metadata](https://datatracker.ietf.org/doc/html/rfc8414)

The discovery flow is illustrated below:  

```mermaid
sequenceDiagram
    participant MC as MCP Client
    participant MS as MCP Server
    participant AS as Authorization Server

    MC->>MS: Request data
    MS-->>MC: 401 Unauthorized with WWW-Authenticate
    MC->>MS: Fetch PRM document
    MS-->>MC: PRM document response
    Note over MC: Validate PRM document<br/>metadata
    Note over MC: Build AS metadata URL
    MC->>AS: Fetch AS metadata
    AS-->>MC: AS metadata response

    Note over MC,AS: Continue with authorization flow
```

#### 2.3.1 Server Metadata Discovery Headers

MCP clients **SHOULD** include the header `MCP-Protocol-Version: <protocol-version>` during Server
Metadata Discovery to allow the MCP server to respond based on the MCP protocol version.

For example:

```http
MCP-Protocol-Version: 2024-11-05
```

### 2.4 Authorization server integration

#### 2.4.1 Overview

MCP servers **MUST NOT** complete public client flows on behalf of the MCP client (e.g., take 

> authorization codes and exchange them for tokens with a public client ID, and then pass the tokens 

Reviewer 12: ...with a public client ID embedded in JavaScript code running in the browser, and then...

Added some details to help getting people familiar with what is public client flow. 

Shall we also mention that "Public client flows **MUST** be completed by MCP Client using Authorization Code Flow with PKCE"? I'm not sure if this is the only recommended auth flow though.

Author: @pwwpche want to submit this as a suggestion so that I can commit on your behalf? I think the latter statement makes sense and complements the statement about clients having to support PKCE.

Reviewer 12: ```suggestion
authorization codes and exchange them for tokens with a public client ID embedded in JavaScript code running in the browser, and then pass the tokens 
```

Author: Ah, I see what you mean here @pwwpche - this is not constrained to just the browser. You can have a server that tries to do public client flows _within the server code_ and it should not be doing that.

Reviewer 12: > Ah, I see what you mean here @pwwpche - this is not constrained to just the browser. You can have a server that tries to do public client flows _within the server code_ and it should not be doing that.

Yeah - browser is just one of the example. I'm not an oauth expert, and when I read the lines above it feels like it needs better clarification to non-experts about what "public client" means. Thus I'm adding "browser" to illustate what public client could mean.

(e.g., take authorization codes and exchange them for tokens with a public client ID (browser, desktop, mobile app, or other client side places that the secret can't be secured), and then pass the tokens)

> back to the client).

Reviewer 12: ```suggestion
back to the client). Public client flows **MUST** be completed by MCP Client using OAuth2 Authorization Code Flow with PKCE
```

Author: Committed this as part of a separate call-out (needed to re-format a bit).

### 1.1 Purpose and Scope

The Model Context Protocol provides authorization capabilities at the transport level, enabling MCP
clients to make requests to protected MCP servers on behalf of resource owners. This specification
defines the authorization flow for HTTP-based transports.

> The scope of this document is **user authorization** - how a client can authenticate a user to access a

Reviewer 12: **MCP** client can authenticate a user to access a **MCP** server,  and the **MCP** server

Adding MCP prefix to keep consistent with the terms in Base tenets section

Author: Updated

server and the server determine whether to authorize the user to access its tools, resources, or any
other capabilities.

This document **does not discuss server-to-server authorization patterns**.

### 1.2 Baseline tenets

1. **MCP server developers are not security experts**. Developers should not be building their own
   authorization servers unless absolutely necessary. This introduces room for serious mistakes
   that can lead to compromise.
1. **MCP clients are Authorization Server (AS)-agnostic**. We do not and will not expect clients to be
   "enlightened" in any capacity to support a protected MCP server. This will fragment the ecosystem.
   The specification explicitly assumes that clients are not enabled in any unique way to support

>    specific Authorization Servers.

Reviewer 12: I've got a bit confused on the "enlightened" part. Trying to rephrase this to make it more understandable:

When you build your MCP server and secure it with an AS, you need to design it so that standard, off-the-shelf MCP clients can connect without needing any special modifications just for your server's specific authentication setup.

Author: Updated.

    MC->>MS: Request data
    MS-->>MC: 401 Unauthorized with WWW-Authenticate
    MC->>MS: Fetch PRM document
    MS-->>MC: PRM document response
    Note over MC: Validate PRM document<br/>metadata
    Note over MC: Build AS metadata URL
    MC->>AS: Fetch AS metadata
    AS-->>MC: AS metadata response

    Note over MC,AS: Continue with authorization flow
```

#### 2.3.1 Server Metadata Discovery Headers

MCP clients **SHOULD** include the header `MCP-Protocol-Version: <protocol-version>` during Server
Metadata Discovery to allow the MCP server to respond based on the MCP protocol version.

For example:

```http
MCP-Protocol-Version: 2024-11-05
```

### 2.4 Authorization server integration

#### 2.4.1 Overview

MCP servers **MUST NOT** complete public client flows on behalf of the MCP client (e.g., take 
authorization codes and exchange them for tokens with a public client ID, and then pass the tokens 
back to the client).

Public client flows **MUST** be completed by MCP Client using OAuth 2.1 Authorization
Code Flow with PKCE.

#### 2.4.2 Flow Description

The third-party authorization flow comprises these steps:

1. MCP client requests data from the MCP server.  

> 1. MCP server responds with `HTTP 401 Unauthorized` and `WWW-Authenticate` header containing the

Reviewer 12: How does MCP Server find out that "the request isn't authorized"? Does it check by token existence?

* If an access token does not exist from the request, then that's easily an unauthorized result. 
* If an access token exists in the request, do we always assume that this token is valid and not expired since MCP Client gives the guarantee?

Author: 1. Yes, that's the clue - if there is no token attached, then that means that the request is unauthorized.
2. Absolutely not - the MCP server should _never_ assume that the token it receives from the client is valid and/or suited for its use. MCP server needs to validate the token through one of the approaches outlined in section `2.9`. The running mantra is "Never trust the client."

Reviewer 2: > If an access token exists in the request, do we always assume that this token is valid and not expired since MCP Client gives the guarantee?

No, the MCP Server (Resource Server) must validate the token it has been presented by the Client. This can happen via one of the many methods available.  See: https://oauth.net/2/bearer-tokens/

Most Authorization Server implementations either issue self-contained tokens (like JWT, etc), or offer some form of token introspection; they usually provide details on how to validate tokens issued by them.  Example [1](https://auth0.com/docs/secure/tokens/access-tokens/validate-access-tokens) [2](https://learn.microsoft.com/en-us/entra/identity-platform/access-tokens) [3](https://developer.okta.com/docs/guides/validate-access-tokens/dotnet/main/).  I do this verification in my prototype sample https://github.com/ciamshrek/mcp-auth-demos/blob/main/mcp-server/src/fastify-auth0-fork.ts#L83-L110

It would then be on the MCP Server to utilize this information. For instance, if the token can include information that grant's the Bearer access to use `read:todos` the MCP Server can determine that its okay to allow calling `read_todo` tool, but not `update_todo` tool.  For [example](https://github.com/ciamshrek/mcp-auth-demos/blob/main/mcp-server/src/tools.ts#L49-L52).

Reviewer 7: There will be challenges for folks that are hosting tools in a MCP Server (mcp.example.com) that wrap/proxy/use a 3P resource server (api.saas.com) that has its own 3P Authorization Server (as.saas.com) that use opaque tokens.  Not all authorization servers allow token introspection for public clients and the MCP Server would need a client relationship with the 3P Authorization Server to have its own client credentials (confidential client) and the Authorization Server would need to allow the MCP Server client to introspect another client's token (MCP Client).   Lots of stars need to align.  Obviously for 1P you can configure your way out this which is what you commonly see in traditional API Gateway deployments.

Author: @mcguinness I agree with the sentiment, but it does feel like some level of token validation enforcement is needed. Otherwise, the client can pass _anything_ to the resource (MCP) server, and without the means to check the validity of the token the server would act on it.

Reviewer 2: > MCP Server (mcp.example.com) that wrap/proxy/use a 3P resource server (api.saas.com) that has its own 3P Authorization Server (as.saas.com) that use opaque tokens.

I am a bit puzzled what you mean here, you mean to say that MCP.example.com is forwarding the token it received to api.saas.com which was originally intended for api.saas.com (why would the client obtain a token like so)?

Reviewer 12: Per @ciamshrek suggested, could we use a fallback approach:

MCP Server **MUST** validate the token
1) If the token is self contained, like is JWT, and MS has obtained the secret for validating JWT, MS must validate the token itself (However, how does MS know that the token is self contained?)
2) If the token is opaque (like Google OAuth token), but AS provides token introspection following [RFC7662](https://datatracker.ietf.org/doc/html/rfc7662), MS **SHOULD** validate the token using the introspect endpoint from AS.
3) If the token is not self contained, and AS does not support token inspection, MCP Server **MUST** leverage any endpoint from AS to test the token validity. (This is the last resort of token validation, and should happen rarely.)

Regarding the case 3) If token is opaque and AS does not support introspection (like in the case @mcguinness commented above)
A MCP Developer in this case is usually developing a MCP that proxies an arbitrary SaaS service. In this case, it is the SaaS service and only the SaaS service that this MCP Server supports, and authenticating to the AS of this SaaS is equal to obtaining access to __all__ tools from this MCP Server. 

In this case, we may assume that the token is valid from MS, and rely completely on the AS of this SaaS for validation. Enforcing token validation from MS would not offer any benefits, but both slow down develop velocity and degrade performance by having an extra token check that AS would eventually do again. @mcguinness Does this align with what you think?

Reviewer 7: > only the SaaS service that this MCP Server supports

I think the only contentious part is whether this is server or tool level.  @wdawsonn or @nbarbettini  may have some thoughts here.

Reviewer 6: > > only the SaaS service that this MCP Server supports
> 
> I think the only contentious part is whether this is server or tool level. @wdawson or @nbarbettini may have some thoughts here.

It's unavoidable pairing the token validation mechanism at the tool level if there are (multiple) resources behind the MCP server protected by different AS right?

Reviewer 3: I don't think we can make any assumptions on how tokens or 3rd party APIs map to services or tools either - couldn't a MCP server expose a single tool that wraps multiple 3rd party APIs together? E.g. a `SearchEmails` tool that accesses all your inboxes across multiple providers. 

I also don't think we can define an all-encompassing relationship between MCP tokens and upstream API credentials - we can recommend popular patterns (reuse token 1-1 when the AS is the same, secure ways to vault tokens when there are multiple AS's) but we can't even assume that _OAuth tokens_ are involved at all. The exact mechanism of MCP -> Upstream API auth is outside the scope of this spec. MCPs could be communicating with a legacy service over FTP for all we know.

Maybe something like

MCP Servers **MUST** take all necessary steps to ensure no data is returned to unauthorized parties.

1. MCP Servers using delegated authorization __MAY__ delegate token validation to the AS or another Resource Server (e.g. using token introspection, JWT validation, or other out-of-band mechanisms outside the scope of this specification)
2. MCP Servers using colocated authorization __MUST__ validate the token before processing the request

Author: @max-stytch on (1), I think that needs to be a **MUST** - essentially, developers of MCP servers always have to validate tokens. You can never just assume that a token is valid. Even if it's not a JWT, it needs some mechanism to ensure that what you're getting is actually a legit credential.

Reviewer 7: Agree @max-stytch that we need this flexibility.  My only feedback is that "colocated authorization" is not a widely used term in OAuth and I think folks mean 1st Party Authorization Server but may be missing the context on why "colocated" is the better term. Whether the AS is hosted in the same "server" or another "server" in the same domain isn't as significant IMHO as whether the Authorization Server and Resource Server share the same administrative and trust boundary and can rely in internal implementation details.

Author: @mcguinness - happy to adjust the terminology, by the way. What would be a better framing here?

Reviewer 3: > essentially, developers of MCP servers always have to validate tokens. You can never just assume that a token is valid. 

The token needs to be validated but I don't think the spec can define precisely how the token gets validated. If a MCP Server wraps an existing REST API and forwards the opaque token to the existing REST API 1-1, does the existing REST API validation count as the MCP Server _also_ validating the token? Does it make sense to mandate the MCP Server locally validate the token in addition to calling out to an external API?

> but may be missing the context on why "colocated" is the better term

I think "colocated" has emerged as the term elsewhere in this spec to describe where a MCP Server is also fulfilling the role of AS and issuing its own tokens, as opposed to delegating the role to an existing AS / SP ecosystem. It doesn't have meaning outside of this spec. 

A meta-q: How much of this document do we need to be strictly specification, and how much of this document is examples? If the spec only said `MCP Servers MUST take all necessary steps to ensure no data is returned to unauthorized parties and MUST ensure the credentials presented are valid before processing the request.` - would that work? That would leave specifics of how a token is validated up to the implementer. Specific examples of how MCP Servers, ASs, SPs, etc. can be deployed could live elsewhere as Guide material. Guide material can move into future versions of the spec once we're more aligned on popular patterns.

Author: I like the framing @max-stytch. I updated the spec to include that, as well as reference examples on token validation.

   URL to the PRM document. 
1. MCP client retrieves and validates the PRM document.
1. MCP client retrieves the AS endpoint from the PRM document.  
1. MCP client redirects user to the AS authorization endpoint. MCP client can include scopes that
   are defined in the PRM document or other scopes, as supported by the MCP server.
1. User authorizes with authorization server.
1. Authorization server redirects back to MCP client with authorization code.  
1. MCP client exchanges the code for the requested tokens.  
1. MCP client stores the token.  

> 1. MCP client can now request data from the MCP server with the token attached in the `Authorization`

Reviewer 12: ```suggestion
1. MCP client can now request data from the MCP server with the token attached in the request, for example in the `Authorization`
```

If MCP client is a public client implemented in the JS code in the browser, client token will be stored and sent to MCP Server via a secured cookie

Author: @pwwpche but is that a pattern we want to document in _this_ spec? Because we haven't really touched on browser-based auth flows here that rely on session/cookies, and this would be different than the rest of the OAuth spec (i.e., this would mean that a lot of other logic could also be different, such as responses to challenges, etc.)

@dsp-ant @aaronpk @freiadr would love your input here.

Reviewer 12: No we don't need to document the cookie one as it's not really a use case right now, just wanna point out that where the token is attached would be affected and restricted by the implementation of the client.

What triggered this comment was MCP Python SDK implementation for SSE. After token is exchanged, the client would actually sends the token through request metadata, instead of in the Authorization header

https://github.com/modelcontextprotocol/python-sdk/blob/main/src/mcp/types.py#L53

Reviewer 13: I would advise against this as well. Authorization headers are the norm in browser/JS scenarios as well. Adding support for cookies would be quite non-standard and requires us to deal with CSRF and other cookie problems.

Reviewer 2: > If MCP client is a public client implemented in the JS code in the browser, client token will be stored and sent to MCP Server via a secured cookie

In all cases the token should be present in the Header, Query or Body. `Authorization` Header is usually preferred.

Author: Let's keep this without references to `cookie`-based flows. Thank you folks for the input!

   header.

```mermaid
sequenceDiagram
    participant UA as User-Agent (Browser)
    participant MC as MCP Client
    participant MS as MCP Server
    participant AS as Authorization Server

    MC->>MS: Request data
    MS-->>MC: 401 Unauthorized with WWW-Authenticate
    Note over MC: MCP Client retrieves and validates<br/>PRM document
    Note over MC: MCP Client retrieves authorization<br/>endpoint from AS metadata document<br/>(Section 2.3)
    MC-->>UA: Initiate authorization request at the authorization endpoint
    Note over UA: User authenticates and authorizes<br/>client for scope and resource
    UA->>MC: Return authorization code to client
    MC->>AS: Exchange code for token
    AS-->>MC: Access token
    MC->>MC: Cache token
```

#### 2.4.4 Security Considerations

When implementing authorization with an existing AS, MCP servers **MUST**:

1. Request the minimum number of required scopes.  

> 2. Securely store any tokens obtained from the client and/or the authorization server.  

Reviewer 12: Could you help elaborate what tokens should be stored in MCP Server? From the Flow Description above I did not found any token that should be stored in the MCP Server, and all tokens are handled by the client.

Author: Here's a scenario - the server can take the user token and then use [token exchange](https://www.rfc-editor.org/rfc/rfc8693.html), or in Microsoft Entra ID parlance, request a [On-behalf-of token](https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-on-behalf-of-flow). It can then cache those server-side without ever exposing them to the client.

Reviewer 12: Oh I see, that's when MCP Server handles token required by its tools (like if a tool needs Microsoft Azure API). Makes sense to me, thanks!

- MCP clients **SHOULD** evaluate or enable the user to evaluate the trustworthiness of authorization
  servers they are connecting to.
- Authorization servers **MUST** support
  [OAuth 2.0 Authorization Server Metadata](https://datatracker.ietf.org/doc/html/rfc8414)

The discovery flow is illustrated below:  

```mermaid
sequenceDiagram
    participant MC as MCP Client
    participant MS as MCP Server
    participant AS as Authorization Server

    MC->>MS: Request data
    MS-->>MC: 401 Unauthorized with WWW-Authenticate
    MC->>MS: Fetch PRM document
    MS-->>MC: PRM document response
    Note over MC: Validate PRM document<br/>metadata
    Note over MC: Build AS metadata URL
    MC->>AS: Fetch AS metadata
    AS-->>MC: AS metadata response

    Note over MC,AS: Continue with authorization flow
```

#### 2.3.1 Server Metadata Discovery Headers

MCP clients **SHOULD** include the header `MCP-Protocol-Version: <protocol-version>` during Server
Metadata Discovery to allow the MCP server to respond based on the MCP protocol version.

For example:

```http
MCP-Protocol-Version: 2024-11-05
```

### 2.4 Authorization server integration

#### 2.4.1 Overview

> MCP servers **MAY** override endpoints on an as-needed basis if the authorization server requires

Reviewer: This paragraph should be removed. We want to ensure there is only one way for the client to discover the authorization server. The AS can be colocated with an RS if the server operator wishes to deploy it in that way, but the discovery mechanism would still be the same as if they were separate.

Author: Updated.

    participant MC as MCP Client
    participant MS as MCP Server
    participant AS as Authorization Server

    MC->>MS: Request data
    MS-->>MC: 401 Unauthorized with WWW-Authenticate
    Note over MC: MCP Client retrieves and validates<br/>PRM document
    Note over MC: MCP Client retrieves authorization<br/>endpoint from AS metadata document<br/>(Section 2.3)
    MC-->>UA: Initiate authorization request at the authorization endpoint
    Note over UA: User authenticates and authorizes<br/>client for scope and resource
    UA->>MC: Return authorization code to client
    MC->>AS: Exchange code for token
    AS-->>MC: Access token
    MC->>MC: Cache token
```

#### 2.4.4 Security Considerations

When implementing authorization with an existing AS, MCP servers **MUST**:

1. Request the minimum number of required scopes.  
2. Securely store any tokens obtained from the client and/or the authorization server.  
3. Implement appropriate session timeout handling.  
4. Consider security implications of token chaining.  
5. Implement proper error handling for authorization failures.

MCP clients **MUST**:

1. Securely store Access and Refresh Tokens received from the AS.  

> 2. Properly handle requests to re-authenticate users with expired or invalid tokens (using the

Reviewer: The refresh token grant is not re-authenticating the user, this should be reworded.

Author: Updated.

Authorization servers **MUST**:

1. Verify that the resource identifier is registered for the audience of the access token issued.

### 2.5 Delegated authorization or colocated AS flows

In limited scenarios, MCP servers **MAY** implement their own AS, refer to as a _colocated AS_ due
to the MCP server _colocating_ itself, the resource server (RS) with the AS.

MCP servers that implement their own AS **MUST** host appropriate AS metadata, as outlined in
[RFC 8414 Section 3](https://datatracker.ietf.org/doc/html/rfc8414#section-3).

MCP servers **MAY** support delegated authorization through third-party AS.
In this flow, the MCP server acts as both an OAuth client (to the external AS) and an OAuth
AS (to the MCP client).

#### 2.5.1 Flow description

The delegated authorization flow comprises these steps:

1. MCP client initiates standard OAuth flow with MCP server
2. MCP server redirects user to third-party authorization server
3. User authorizes with third-party server
4. Third-party server redirects back to MCP server with authorization code
5. MCP server exchanges code for third-party access token
6. MCP server generates its own access token bound to the third-party session

> 7. MCP server completes original OAuth flow with MCP client

Reviewer 12: ```suggestion
7. MCP server returns its own access token to the MCP Client
8. MCP Server completes original OAuth flow with MCP client
```

Author: I think that's encapsulated in the last step (completing the OAuth flow), which means that the server returns issues token to the client. Maybe rephrase that last point?

Reviewer 12: The reason that should be explicitly mentioned is due to there being two access tokens - third-party access token and MS access token, so it's better to clarify which MS should return.

I tend to get very detailed in the docs, feel free to take your own judgement if it's a bit too detailed :)

Author: Resolving because the original diagram is no longer included.

```mermaid
sequenceDiagram
    participant UA as User-Agent (Browser)
    participant MC as MCP Client
    participant MS as MCP Server
    participant T as Third-Party Auth Server

    MC->>MS: Initial OAuth Request

>     MS->>UA: Redirect to Third-Party /authorize

Reviewer 12: ```suggestion
    MS->>MC: Generates redirection url to AS Authorization Endpoint
    MC->>UA: Redirect user to Third-Party /authorize
```

MS shouldn't directly talk to UA.

### 1.1 Purpose and Scope

The Model Context Protocol provides authorization capabilities at the transport level, enabling MCP
clients to make requests to protected MCP servers on behalf of resource owners. This specification
defines the authorization flow for HTTP-based transports.

The scope of this document is **user authorization** - how a MCP client can authenticate a user
to access a MCP server and the MCP server determine whether to authorize the user to access its tools,
resources, or any other capabilities.

This document **does not discuss server-to-server authorization patterns**.

### 1.2 Baseline Tenets

> 1. **MCP server developers are not security experts**. Developers should not be building their own

Reviewer 14: This statement seems to sensationalize, and is not quite true. I'd prefer being a little less forceful about who MCP server developers are.

Reviewer 15: This statement is there on the heels of quite a few customer interviews and discussions we had prior to writing this RFC 😄

We can re-frame it ("Most MCP server developers are not security experts") but I think it's fair to say that the vast majority of developers who want to expose MCP servers should not be re-implementing security capabilities from scratch - because security is not their wheelhouse (nor do they want it to be).

Reviewer 16: IMO this can be framed in a way that does not touch on what fraction of MCP devs are security experts, because honestly it is kind of irrelevant since this won't be a constant over time depending on how the software ecosystem evolves. The more important question is whether they are expected to be security experts, and the answer there is clearly no.

ie. 
> **MCP server developers are not expected to be security experts**.

Author: Good feedback - I've amended the spec @allenzhou101 @blowdart @juehang

   ([RFC 7591](https://datatracker.ietf.org/doc/html/rfc7591)).

3. Authorization servers **SHOULD** implement and MCP clients **MUST** consume OAuth 2.0 Authorization
   Server Metadata ([RFC 8414](https://datatracker.ietf.org/doc/html/rfc8414)).

4. MCP servers **MUST** implement and support Auth 2.0 Protected Resource Metadata
   ([RFC DRAFT](https://datatracker.ietf.org/doc/draft-ietf-oauth-resource-metadata/13/)).

### 2.2 Client OAuth 2.1 Authorization

When authorization is required and not yet proven by the client, servers **MUST** respond with
`HTTP 401 Unauthorized`.

A `HTTP 401 Unauthorized` response **MUST** include a `WWW-Authenticate` header with
the URL to the protected resource metadata document hosted by the MCP Server in
the following format:

```http
WWW-Authenticate: Bearer resource_metadata="https://resource.example.com/.well-known/oauth-protected-resource",
     scope="mcp:read mcp:write"
```

A MCP server **MUST** host its Protected Resource Metadata (PRM) document, as outlined in
[Section 3 of the RFC draft](https://www.ietf.org/archive/id/draft-ietf-oauth-resource-metadata-13.html#section-3).

> A MCP server **MUST** return a set of scopes in the `scope` parameter within the `WWW-Authenticate` header

Author: That's a good call-out - I would support the idea of having the _minimal required_ set of scopes, and then do incremental changes with subsequent 401s.

in the initial handshake as well as in subsequent responses that require incremental authorization.

A MCP server **MUST** include its AS configuration in the PRM document as outlined in
[Section 2 of the RFC draft](https://www.ietf.org/archive/id/draft-ietf-oauth-resource-metadata-13.html#section-2).

A MCP client **MUST** issue a `GET` request to the URL in `resource_metadata` to get information about
the AS.

> A MCP client **MUST** verify that the `resource` value in the PRM document matches the MCP server

Reviewer 17: Resouce values include a protocol as well as the FQDN. Should the protocol also be matched?

Author: Good catch - updated.

### 1.1 Purpose and Scope

The Model Context Protocol provides authorization capabilities at the transport level, enabling MCP
clients to make requests to protected MCP servers on behalf of resource owners. This specification
defines the authorization flow for HTTP-based transports.

> The scope of this document is **user authorization** - how a MCP client can authenticate a user

Reviewer 18: I have to say I am confused reading this right at the start. Apart from the authN/authZ mingling (which one is used to nowadays), OAuth is NOT user authorization. ever. It is an authorization protocol that allows a client application  to request permission to access a protected resource  on the resource owner’s behalf [[1](https://www.scottbrady.io/oauth/oauth-is-not-user-authorization)] - or, as per the specs abstract: It 

> ... enables a third-party application to obtain limited access to an HTTP service, either on  behalf of a resource owner by orchestrating an approval interaction between the resource owner and the HTTP service, or by allowing the  third-party application to obtain access on its own behalf...
[[2](https://datatracker.ietf.org/doc/html/rfc6749)]

What are we talking about here? User authZ oder delegated client authZ? 
I think the latter, but read here (and e.g. in some diagrams in 2.58.1) the former. 

If I may have one wish: Please avoid this confusion in a spec, where words really matter. When we talk about user authZ, I have multiple thoughts and ideas on how ReBAC- or ABAC-Systems and Drafts such as [openID authZEN](https://openid.net/wg/authzen/charter/)* could be a starting point to simplify this whole complicatedness a lot and solve multiple problems like the confused deputy on the way. But I think this spec is not about that topic... right?

*: Subjective thought: The authZEN spec is the right direction, XACML is not, and should be deleted from the authZEN implementers draft - but what do I know *shrugs

Reviewer 19: I beg to differ a little @DGuhr  -- from the client's point of view, the access token represents what the user has authorized the client to do => OAuth is how a client get's user authorization to access the user's resources -- the request from the client is represented by the `scope` in the flow. Hence the user authorization is scoped to what the user authorized the client to do with their resources. 

This user authorization is a subset of what the user is authorized to do. 

I agree that words are important and the term "user authorization" means different things depending on the perspective being taken. 

From the user's point of view, it is what the user has been authorized to do -- but in this document, it is what the user has authorized the client to do. Make sense?

wrt. scopes -- fine grained access is impractical to be represented to a user in an OAuth flow for them to understand what is being granted. I bet that fine grained access is only needed at the user level, not at the client level.

Reviewer 18: @dickhardt What you're describing here: 
> [...] from the client's point of view, the access token represents what the user has authorized the client to do => OAuth is how a client get's user authorization to access the user's resources -- the request from the client is represented by the scope in the flow. Hence the user authorization is scoped to what the user authorized the client to do with their resources.
[...]
This user authorization is a subset of what the user is authorized to do.

is exactly the concept of "authorization on behalf of" / delegated access as of my understanding.  As such, the appropriate wording made it into several IETF-Specs, because there is a difference and you tap into all kinds of problem classes when OAuth is used for user authZ, as it is inherently not made for it. 

I just scanned all the IETF RFCs linked in this spec, and also RFC 6479.  The latter is actually the only spec that mentions user authorization (searched for "user authoriz") at all, but only in two sentences where from the context  it is clear that  clients are meant (e.g. "... end-user authorization for a malicious client").

I am just saying the used wording creates unnecessary ambiguity, that likely leads to wrong usage of a spec family that was not designed to be used for this kind of authZ. I'll give you that OAuth _is_ already widely used for it, with all kinds of resulting problems in the real world. See e.g. the linked article, or ask your llm of choice "where's the problem using OAuth for user authorization?" - I just did, the result was good enough. 

In the end all I say is the wording confused me, as it would likely confuse others. I'd wish for it to be changed to adhere to IETF-used wording. Especially, as implementing the (remote) MCP spec w.r.t. authZ already seems to be non-trivial at all for most, which could hurt adoption badly. I personally do not see many customers using varying audiences - though they should. Or DPoP - which was a nice idea (token binding), but where the newest incarnation renders it also open to varying [attack vectors](https://github.com/w3c/webappsec-dbsc/issues/136). Let alone dynamic client registration and more. In the world that I live in, even ROPC is still used often in 2025. 

I think what I want to say is: On top of all the complicated things that already made it into the spec, people shouldn't need to figure out what is meant here with "user authorization". 

> wrt. scopes -- fine grained access is impractical to be represented to a user in an OAuth flow for them to understand what is being granted. I bet that fine grained access is only needed at the user level, not at the client level.

I think I'm with you here for client authZ, though I'd "not bet on it", given the "generic container for basically everything"-nature of MCP Servers. Time will tell.

to access a MCP server and the MCP server determine whether to authorize the user to access its tools,
resources, or any other capabilities.

> This document **does not discuss server-to-server authorization patterns**.

Reviewer 20: This explicit scope is great, and I also think it's more reasonable to limit the OAuth2 related specifications described in the document to the **user authorization** scenario. However, the description in section 2.2.1 here is inconsistent with this scope agreement:

> Client Credentials: the client is another application (not a human)

Should the description in section 2.2.1 be removed? I believe Client Credentials should be considered in server-to-server authorization patterns.

Author: That's a fair call-out. I think the grant section alludes to the fact that _different grants_ may be supported between clients and servers. Ultimately, a client may choose to add support for client credentials too (as impractical as that may be for the general case), but this specification does not cover the implementation or architectural implications.

   them with an AS, they need to design them so that standard, off-the-shelf MCP clients can connect
   without needing any special modifications for the server's authorization setup.
   Not doing this will fragment the ecosystem. The specification explicitly assumes that clients are
   not enabled in any unique way to support specific AS implementations beyond what is covered here.
    - **Note:** It is acceptable for clients to be enhanced for specific Authorization Servers,
      primarily through pre-registration.
1. **User Authentication is not a MCP Server concern.** The MCP server can guide the client
   to the AS but will not be responsible for authenticating the user and obtaining credential
   artifacts for the user.
1. **We focus on OAuth 2.1, with room for extensibility for other providers/specifications.** Our
   guiding assumption is that most developers will want to use OAuth 2.1, however in the long-run
   unusual cases for authentication and authorization may emerge (e.g., mTLS, Kerberos), which will
   need clients to be extensible.

### 1.3 Protocol Requirements

Authorization is **OPTIONAL** for MCP implementations. When supported:

> - Streamable HTTP or SSE transports **SHOULD** conform to this specification.

Reviewer 21: this leaves out websockets, which would also want to handle this. The former wording was more permissive

   (see [RFC 9700](https://datatracker.ietf.org/doc/html/rfc9700)) for both confidential and
   public clients.

2. Authorization servers **SHOULD** support the OAuth 2.1 Dynamic Client Registration Protocol
   ([RFC 7591](https://datatracker.ietf.org/doc/html/rfc7591)).

3. Authorization servers **SHOULD** implement and MCP clients **MUST** consume OAuth 2.0 Authorization
   Server Metadata ([RFC 8414](https://datatracker.ietf.org/doc/html/rfc8414)).

4. MCP servers **MUST** implement and support Auth 2.0 Protected Resource Metadata
   ([RFC DRAFT](https://datatracker.ietf.org/doc/draft-ietf-oauth-resource-metadata/13/)).

### 2.2 Client OAuth 2.1 Authorization

When authorization is required and not yet proven by the client, servers **MUST** respond with
`HTTP 401 Unauthorized`.

A `HTTP 401 Unauthorized` response **MUST** include a `WWW-Authenticate` header with
the URL to the protected resource metadata document hosted by the MCP Server in
the following format:

```http
WWW-Authenticate: Bearer resource_metadata="https://resource.example.com/.well-known/oauth-protected-resource",
     scope="mcp:read mcp:write"

> ```

Reviewer 22: What about the scenarios where a new authorization scope is needed as part of responding/steaming?

I imagien this is good for initial handshake and such, but what if client is doign a toolcall where the MCP server finds out that it requires additional authorization to continue.  I would be great if the spec covers a standalized way tool calls should response with needed authorization similar to www-authenticate over http does? 

Else we end up with servers and clients that are made specially to understand eachother and not easy to swap to a new client.

```
    let hasValidUpstreamToken = await hasValidToken(azureAdProvider.id, userId);

    if ('error' in hasValidUpstreamToken) {
      return { error: "No valid upstream provider tokens available. Authorize the mcp://azure-resoruce-manager scope.\nDetails:"+ hasValidUpstreamToken.error };
    }
```

Author: I'll say that tool authorization is a separate issue from this spec - as @max-stytch called out, there are questions about things like incremental consent that need to be addressed, but that would fall somewhat out of scope of this particular RFC.

@nbarbettini and @wdawson have been giving a lot of thought to the tool authorization problem and are looking at proposing a few tweaks in the future. Let's take that discussing there when it's ready.

that require incremental authorization.

A MCP server **MUST** include its AS configuration in the PRM document as outlined in
[Section 2 of the RFC draft](https://www.ietf.org/archive/id/draft-ietf-oauth-resource-metadata-13.html#section-2).

A MCP client **MUST** issue a `GET` request to the URL in `resource_metadata` to get information about
the AS.

A MCP client **MUST** verify that the `resource` value in the PRM document matches the MCP server
protocol and fully-qualified domain name (FQDN).

A MCP client **SHOULD** evaluate or permit the user to evaluate the trustworthiness of AS
discovered through the PRM document to prevent phishing. For example, an MCP client could prompt 
the user "Do you want to login with `https://authserver.example`?"

A MCP client **MUST** use `authorization_servers` values from the PRM document to construct
the OAuth 2.0 metadata discovery document path, following
[RFC 8414](https://datatracker.ietf.org/doc/html/rfc8414#section-3).

A MCP client **MAY** use the AS-provided registration endpoint to register itself with the AS.

A MCP client **MAY** be pre-registered with an AS and not require dynamic registration.

> A MCP client **MUST** send the `resource` identifier to the AS, following

Reviewer 23: Is there an implicit requirement that the AS should support and implement RFC 8707? There is another point in section 2.4.4 which also implies that. If so, this could be difficult to enforce given that the AS could audience-restrict the token using a different strategy and the outcome would be the same.

[RFC 8707](https://www.rfc-editor.org/rfc/rfc8707.html).

A MCP client **MAY** implement Demonstrating Proof of Possession (DPoP) and use it to the extent
supported by the AS. See
[RFC 9449](https://datatracker.ietf.org/doc/html/rfc9449).

> A MCP clients **MUST** initiate an

Reviewer 24: MUST seems strong here. Shouldn't MCP clients be allowed to treat the data as optional and proceed without it?

Author: This spec is covering protected MCP servers, so if a client receives a `401 Unauthorized`, what should it then do other than re-initiate the authorization flow?

Reviewer 24: If the user granted only partial consent on a previous authorization flow, the client shouldn't badger the user for consent unless the data is essential to the task. It's increasingly likely in a world with LLMs that data access only improves result quality rather than actually being essential. (e.g. "find me a good Italian recipe for tonight" would be a lot better with access to a bunch of my personal data, but it doesn't _need_ it)

Author: Sure, but the server returns a `401 Unauthorized` - should the client just ignore it?

Reviewer 21: The clients can actually be smart enough to make that decision, that's the beauty of having an LLM. I would say should is more appropriate

Reviewer 24: @localden it's not that they _should_ ignore it but they should be allowed to if they want. I agree with Pat that we should either soften the language to SHOULD or say "MUST initiate ... if the data access is required".

Reviewer 21: Another good example - a local security policy could explicitly forbid a particular auth target, there’s no reason that would cause them to be non compliant with the spec

[OAuth 2.1 (draft)](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12) authorization
flow after receiving a `HTTP 401 Unauthorized` response.

### 2.2.1 OAuth Grant Types

OAuth specifies different flows or grant types, which are different ways of obtaining an

> access token. Each of these targets different use cases and scenarios.

Reviewer 23: This is intended for the section below, which I can't tie my comment to.
> MCP servers **SHOULD** support the OAuth grant types that best align with the intended
audience. For instance:

Is this still valid given that MCP servers are now resource servers, which are generally not concerned with the authorization grant type?

- MCP clients **SHOULD** evaluate or enable the user to evaluate the trustworthiness of authorization
  servers they are connecting to.
- Authorization servers **MUST** support
  [OAuth 2.0 Authorization Server Metadata](https://datatracker.ietf.org/doc/html/rfc8414)

The discovery flow is illustrated below:  

```mermaid
sequenceDiagram
    participant MC as MCP Client
    participant MS as MCP Server
    participant AS as Authorization Server

    MC->>MS: Request data
    MS-->>MC: 401 Unauthorized with WWW-Authenticate
    MC->>MS: Fetch PRM document
    MS-->>MC: PRM document response
    Note over MC: Validate PRM document<br/>metadata
    Note over MC: Build AS metadata URL
    MC->>AS: Fetch AS metadata
    AS-->>MC: AS metadata response

    Note over MC,AS: Continue with authorization flow
```

#### 2.3.1 Server Metadata Discovery Headers

MCP clients **SHOULD** include the header `MCP-Protocol-Version: <protocol-version>` during Server
Metadata Discovery to allow the MCP server to respond based on the MCP protocol version.

For example:

```http
MCP-Protocol-Version: 2024-11-05
```

### 2.4 Authorization Server Integration

#### 2.4.1 Overview

> MCP clients **SHOULD** be considered _public clients_.

Reviewer 25: +1 both confidential and public clients should be allowed

> MCP clients **SHOULD** be considered _public clients_.

Author: This would hold if we assume that the client is "no human in the loop" component. I do wonder, however, if we need to draw that distinction _somehow_ in the spec. If we're talking about the _client credentials_ grant, that would almost be a server-to-server (S2S) scenario. Is the "client" really then just another "server"?

This spec is _mostly_ covering the "human in the loop - how do we authorize their client" flow, though.

> MCP clients **SHOULD** be considered _public clients_.

Reviewer 25: i think what is being referred to above is why not allow confidential clients which can guarantee keeping client credentials safe as opposed to starting with public clients

> MCP clients **SHOULD** be considered _public clients_.

Author: I am not sure I entirely understand your point @ishan-chawla-glean - this is marked as **SHOULD** because this is a fairly typical pattern for human-in-the-loop scenarios. An app that you install on your phone or desktop or say a SPA application can be a public client. This is what the spec mostly talks about. If the developer wants to implement a confidential client app, they can. There is no inherent security vulnerability/risk in starting "human-in-the-loop clients should be public clients where possible."

   header.

```mermaid
sequenceDiagram
    participant UA as User-Agent (Browser)
    participant MC as MCP Client
    participant MS as MCP Server
    participant AS as Authorization Server

    MC->>MS: Request data
    MS-->>MC: 401 Unauthorized with WWW-Authenticate
    Note over MC: MCP Client retrieves and validates<br/>PRM document
    Note over MC: MCP Client retrieves authorization<br/>endpoint from AS metadata document<br/>(Section 2.3)
    MC-->>UA: Initiate authorization request at the authorization endpoint
    Note over UA: User authenticates and authorizes<br/>client for scope and resource
    UA->>MC: Return authorization code to client
    MC->>AS: Exchange code for token
    AS-->>MC: Access token
    MC->>MC: Cache token
    MC->>MS: Request data</br>(with Authorization header)
```

#### 2.4.4 Security Considerations

When implementing authorization with an existing AS, MCP servers **MUST**:

1. Request the minimum number of required scopes.  
2. Securely handle any tokens that need to be stored or otherwise managed by the MCP server.  
3. Implement appropriate session timeout handling.  

> 4. Consider security implications of token chaining.  

Reviewer 17: This is "fluffy". Developers need to know what the security implications are :)

Reviewer 2: Maybe link the BCP here @localden

   [RFC 9470 OAuth 2.0 Step Up Authentication Challenge Protocol](https://www.rfc-editor.org/rfc/rfc9470.html).

### 2.7 Dynamic Client Registration

MCP clients **SHOULD** support the
[OAuth 2.0 Dynamic Client Registration Protocol](https://datatracker.ietf.org/doc/html/rfc7591) to
allow MCP clients to obtain OAuth client IDs without user interaction. This provides a standardized
way for clients to automatically register with new servers, which is crucial for MCP because:

- Clients cannot know all possible servers in advance  
- Manual registration would create friction for users  
- It enables seamless connection to new servers  
- Servers can implement their own registration policies

MCP clients that want to dynamically register OAuth clients with an authorization server **MUST**
provide the `software_id` and `software_version` parameters, as defined in the
[Client Metadata](https://datatracker.ietf.org/doc/html/rfc7591#section-2) section of
[RFC 7591](https://datatracker.ietf.org/doc/html/rfc7591).

MCP clients with mass audiences will *generally* want to pre-register and hardcode client IDs.

> Some authorization servers (such as Google and Microsoft) limit the functionality of dynamically

Author: I think that is out-of-scope for this specification, but I agree that it requires more thinking. There is also work from @aaronpk with [OAuth Client ID Metadata Document](https://datatracker.ietf.org/doc/draft-parecki-oauth-client-id-metadata-document/) to potentially help shape this better _without_ DCR.

  servers they are connecting to.
- Authorization servers **MUST** support
  [OAuth 2.0 Authorization Server Metadata](https://datatracker.ietf.org/doc/html/rfc8414)

The discovery flow is illustrated below:  

```mermaid
sequenceDiagram
    participant MC as MCP Client
    participant MS as MCP Server
    participant AS as Authorization Server

    MC->>MS: Request data
    MS-->>MC: 401 Unauthorized with WWW-Authenticate
    MC->>MS: Fetch PRM document
    MS-->>MC: PRM document response
    Note over MC: Validate PRM document<br/>metadata
    Note over MC: Build AS metadata URL
    MC->>AS: Fetch AS metadata
    AS-->>MC: AS metadata response

    Note over MC,AS: Continue with authorization flow
```

#### 2.3.1 Server Metadata Discovery Headers

MCP clients **SHOULD** include the header `MCP-Protocol-Version: <protocol-version>` during Server
Metadata Discovery to allow the MCP server to respond based on the MCP protocol version.

For example:

```http
MCP-Protocol-Version: 2024-11-05
```

### 2.4 Authorization server integration

#### 2.4.1 Overview

MCP clients **SHOULD** be considered _public clients_ and MCP servers **MUST** be

> considered _confidential clients_, as defined in

Author: I actually think that part is good where it is because it outlines the requirement for integration. MCP clients should not store any secrets within them (and if they do - assume that those will be exfiltrated) - I know that some providers don't really support public client flows, so for clients that seems like a "You know what you're getting into" statement.

For servers, [you can do PKCE through confidential clients](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics#section-2.1.1-2.2) too. What would be the scenario where the MCP server needs to operate as a public client?

> considered _confidential clients_, as defined in

Reviewer 3: That makes sense. Calling out the requirements for MCP clients here makes sense to me!

> MCP servers **MUST** be considered _confidential clients_

This is the part I think should be under 2.5 - in the regular flow, the MCP server is not an OAuth client at all. The MCP server is only an OAuth client in the delegated OAuth scenario.

> considered _confidential clients_, as defined in

Author: Ah, yes - I see your point @max-stytch. Updated!

1. MCP client exchanges the code for the requested tokens.
1. MCP client stores the token.
1. MCP client can now request data from the MCP server with the token attached in the `Authorization`
   header.

```mermaid
sequenceDiagram
    participant UA as User-Agent (Browser)
    participant MC as MCP Client
    participant MS as MCP Server
    participant AS as Authorization Server

    MC->>MS: Request data
    MS-->>MC: 401 Unauthorized with WWW-Authenticate
    Note over MC: MCP Client retrieves and validates<br/>PRM document
    Note over MC: MCP Client retrieves authorization<br/>endpoint from AS metadata document<br/>(Section 2.3)
    MC-->>UA: Initiate authorization request at the authorization endpoint
    Note over UA: User authenticates and authorizes<br/>client for scope and resource
    UA->>MC: Return authorization code to client
    MC->>AS: Exchange code for token
    AS-->>MC: Access token
    MC->>MC: Cache token
    MC->>MS: Request data</br>(with Authorization header)
```

#### 2.4.4 Security Considerations

When implementing authorization with an existing AS, MCP servers **MUST**:

1. Request the minimum number of required scopes.  

> 2. Securely store any tokens obtained from the client and/or the authorization server.  

Reviewer 3: Sorry to kick this one open again - the phrasing here makes me think that the MCP server is required to store the credentials from the client, but that shouldn't be a requirement.

Would this line be better under `2.5.1`? In the standard delegated AS model the MCP server can be made stateless - acquire token from request; validate token; return response. Only with delegated OAuth does the MCPS retrieve a token directly from the authorization server.

The token exchange example provided previously is a server-to-server authorization pattern and outside the scope of this spec.

Author: I reframed this a bit to be more generic and less prescriptive.

to the MCP server _colocating_ itself, the resource server (RS) with the AS.

MCP servers that implement their own AS **MUST** host appropriate AS metadata, as outlined in
[RFC 8414 Section 3](https://datatracker.ietf.org/doc/html/rfc8414#section-3).

MCP servers **MAY** support delegated authorization through third-party AS.
In this flow, the MCP server acts as both an OAuth client (to the external AS) and an OAuth
AS (to the MCP client).

#### 2.5.1 Special considerations

MCP servers that use delegated authorization or implement a colocated AS **MUST**:

1. Maintain secure mapping between third-party tokens and issued client sessions.
1. Never pass third-party tokens directly to the client.
1. Validate token status before granting any data access.
1. Implement appropriate token lifecycle management.  
1. Provide appropriate responses to the client if re-authentication is needed. Please refer to
   [RFC 9470 OAuth 2.0 Step Up Authentication Challenge Protocol](https://www.rfc-editor.org/rfc/rfc9470.html).

### 2.6 Implementing authorization server overrides

In certain scenarios, AS might have limited capability support, such as no dynamic client registration, and

> MCP server developers may choose to implement those capabilities themselves.

Author: Yep, good catch - I removed that. I think we can safely state that if you implement your AS, you have to do all AS-related things.

    participant MC as MCP Client
    participant MS as MCP Server
    participant AS as Authorization Server

    MC->>MS: Request data
    MS-->>MC: 401 Unauthorized with WWW-Authenticate

    MC->>MS: Fetch PRM document
    MS-->>MC: PRM document response
    Note over MC: Validate PRM document<br/>metadata
    Note over MC: Build AS metadata URL
    MC->>AS: Fetch AS metadata
    AS-->>MC: AS metadata response

    alt External authorization server
        MC->>AS: GET /.well-known/oauth-authorization-server
        AS-->>MC: 200 OK + Metadata Document
        Note over MC: Use endpoints from metadata
    else Co-located authorization server
        MC->>MS: GET /.well-known/oauth-authorization-server
        MS-->>MC: 200 OK + Metadata Document
        Note over MC: Use endpoints from metadata
    end

    alt Client is pre-registered with the authorization server
    else Authorization server supports dynamic client registration
        MC->>AS: Request client registration
        AS-->>MC: Confirm client registration
    else Authorization server does not support dynamic client registration
        MC->>MS: Request client registration

>         MS->>AS: Register client

Reviewer 3: This step confuses me - the upstream AS does not supports DCR but the MS does? Wouldn't this make the MS the AS under the delegated auth case?

I think this section should instead show the `Alternatively, an MCP client could allow the user to register the MCP client and provide those details to the MCP client through a UI.` state described on line 359
```
MC->> UA: Request Out-of-band Client Registration
UA->>UA: Complete Out-of-band Client Registration
UA-->>MC: Return client registration 
```

Author: In this case, the MS could request client registration with the AS through exposed APIs, even if DCR is not supported directly (e.g., through AS-specific endpoints, like [Graph for Entra ID](https://learn.microsoft.com/en-us/graph/api/application-post-applications?view=graph-rest-1.0&tabs=http)). This is something I've also discussed with @allenzhou101 before in the context of Descope, where DCR might not be supported directly, but the workaround is to wrap AS registration APIs in a custom endpoint.

But you're right @max-stytch that at that point you become the AS (even if you redirect parts to your own, parts to third-party AS).

2. Securely handle any tokens that need to be stored or otherwise managed by the MCP server.  
3. Implement appropriate session timeout handling.  
4. Consider security implications of token chaining, such as ensuring that the appropriate
   tokens are exposed to the right audiences (e.g., don't pass server-issued tokens to the
   client), validating the original scopes and issuers, and looking for unintended privilege
   escalation by virtue of token exchange. Refer to the [Best Practices](#3-best-practices)
   section for additional recommendations.
5. Implement proper error handling for authorization failures.

MCP clients **MUST**:

1. Securely store Access and Refresh Tokens received from the AS.  
2. Properly handle requests to renew and replace expired or invalid tokens (using the
   Refresh Token grant) or due to challenges (refer to
   [RFC 9470](https://www.rfc-editor.org/rfc/rfc9470.html)).
3. Verify that the `resource` identifier in the PRM document matches the MCP server.

Authorization servers **MUST**:

1. Verify that the resource identifier is registered for the audience of the access token issued.

### 2.5 Delegated Authorization or Colocated AS

> In limited scenarios, MCP servers **MAY** implement their own AS, referred to as a _colocated AS_ due

Reviewer 24: What does "limited" mean here?

Maybe rephrase as "MCP servers MAY implement their own AS ... though this should be limited to scenarios where there is significant authorization expertise amongst the MCP server owners."

or something of that nature.

Author: Great feedback. Updated.

MCP servers **MUST** be considered _confidential clients_, as defined in
[RFC 6749 Section 2.1](https://datatracker.ietf.org/doc/html/rfc6749#section-2.1).

MCP servers that implement their own AS **MUST** host appropriate AS metadata, as outlined in
[RFC 8414 Section 3](https://datatracker.ietf.org/doc/html/rfc8414#section-3).

MCP servers **MAY** support delegated authorization through third-party AS.
In this flow, the MCP server acts as both an OAuth client (to the external AS) and an OAuth
AS (to the MCP client).

#### 2.5.1 Special Considerations

MCP servers that use delegated authorization or implement a colocated AS **MUST**:

1. Maintain secure mapping between third-party tokens and issued client sessions.
1. Never pass third-party tokens directly to the client.
1. Validate token status before granting any data access.
1. Implement appropriate token lifecycle management.  
1. Provide appropriate responses to the client if re-authentication is needed. Please refer to
   [RFC 9470 OAuth 2.0 Step Up Authentication Challenge Protocol](https://www.rfc-editor.org/rfc/rfc9470.html).

### 2.7 Dynamic Client Registration

> MCP clients **SHOULD** support the

Reviewer 24: If the spec is going to use SHOULD here (and "strongly recommend" below), I think it's worth an extra callout of the security considerations in RFC7591 around rogue clients attempting impersonation. Given the amount of data that LLMs may attempt to access, the risk is higher and mistakes costlier.

Author: Great call-out, updated.

[OAuth 2.0 Dynamic Client Registration Protocol](https://datatracker.ietf.org/doc/html/rfc7591) to
allow MCP clients to obtain OAuth client IDs without user interaction. This provides a standardized
way for clients to automatically register with new servers, which is crucial for MCP because:

- Clients cannot know all possible servers in advance  

> - Manual registration would create friction for users  

Reviewer 24: Should say "friction for developers"? Users are only seeing the final product anyway.

Author: It's both, because in some alternative view there is the possibility that client users _can_ provide the client ID they want to use with the server (as rough as that approach is from a usability perspective). I updated the verbiage for it.

   - For instance, an agent calls a secure MCP tool to check inventory at a specific
     store. No need to impersonate the end user.

> MCP clients **MUST** implement PKCE when using the authorization code flow, per OAuth 2.1 specification."

Reviewer 26: This quotation mark should probably not be there

```suggestion
MCP clients **MUST** implement PKCE when using the authorization code flow, per OAuth 2.1 specification.
```

    I --> J

    J --> K[Generate PKCE parameters]
    K --> L[Request authorization]
    L --> M[User authorization]
    M --> N[Exchange code for tokens]
    N --> O[Use access token]
```

### 2.9 Access Token Usage

#### 2.9.1 Token Requirements

Access token handling **MUST** conform to
[OAuth 2.1 Section 5](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12#section-5) or
[DPoP Section 7](https://datatracker.ietf.org/doc/html/rfc9449#section-7) (if client
supports DPoP) requirements for resource requests.

MCP client **MUST** use the `Authorization` request header field and **MUST NOT** use URI query string 
to transport tokens [OAuth Section 5.1.1](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12#section-5.1.1):

```http
Authorization: Bearer <access-token>
```

or
[DPoP Section 7.1](https://datatracker.ietf.org/doc/html/rfc9449#name-the-dpop-authentication-sch)
(if client supports DPoP):

```http
Authorization: DPoP <access-token>
DPoP: <DPoP proof>
```

Note that `Authorization` header **MUST** be included in every HTTP request from client to server,
even if they are part of the same logical session.

#### 2.9.2 Token Handling

> Resource servers (MCP servers _or_ downstream APIs) **MUST** validate access tokens as described

Reviewer 27: Consider referencing the token validation in the Security Considerations section later in this document, which outlines critical aspects of token validation.

in [Section 5.2](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12#section-5.2).

If validation fails, servers **MUST** respond according to
[Section 5.3](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12#section-5.3) error
handling requirements. Invalid or expired tokens **MUST** receive a `HTTP 401` response.

### 2.10 Security Considerations

The following security requirements **MUST** be implemented:

1. MCP clients **MUST** securely store tokens following OAuth 2.1 best practices, as
   outlined in the
   [OAuth 2.1 Authorization Framework](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12).
1. MCP servers **MUST** take all necessary steps to ensure no data is returned to unauthorized parties
   and **MUST** ensure any credentials are valid before processing the request. For example, a MCP server
   could validate inbound tokens through one of the following approaches:

>   1. Token introspection, according to [RFC 7662](https://datatracker.ietf.org/doc/html/rfc7662).

Reviewer 27: When I look at the rendered version of this markdown, it appears that this subsection of the list is incorrectly rendered, as it doesn't seem to have the correct indentation. It could be a GitHub preview rendering issue though, don't know.

   - For instance, an agent calls a secure MCP tool to check inventory at a specific
     store. No need to impersonate the end user.

> MCP clients **MUST** support authorization code flow with PKCE, per
> OAuth 2.1 specification.

Reviewer 23: Would this be better worded as:
```suggestion
MCP clients **MUST** implement PKCE when using the authorization code flow, per OAuth 2.1 specification."
```
since clients do not "support" flows.

MCP clients **MUST** present valid tokens with the protected MCP server as the audience and **MUST NOT**
use the MCP server as a pass-through to channel tokens issued to different audiences through it.

Protected MCP servers that receive tokens with an audience different than itself **MUST** reject
them under all circumstances (see [Section 3.11](#210-security-considerations) for validation
considerations).

MCP clients are responsible for performing an OAuth authorization flow, as outlined in
[Section 2.2.1](#221-oauth-grant-types) against the AS declared in the MCP server's PRM document.
This is the only OAuth relationship the client participates in.

Downstream OAuth calls (e.g., to third-party APIs) are implementation details left to the MCP
server developers to design. For those scenarios, the MCP server itself becomes an OAuth client,
and downstream interactions are handled entirely by the MCP server. The MCP client's authorization
relationship remains solely between itself and the MCP server's declared AS.

### 3.5 Third-Party API Access

The third-party API connection process is done out-of-band and does not depend on the authorization
flow outlined in this specification.

> If the MCP server connects to third-party APIs, the MCP client **MUST NOT** hold tokens for said

Author: I think this is more of a reason for a specific _API connection flow_ than expose that through the server authorization specification. Those are fundamentally two different concepts, IMO. One is "I am protecting my server from unauthorized access" and the other is "My server needs to access upstream APIs."

> If the MCP server connects to third-party APIs, the MCP client **MUST NOT** hold tokens for said

Reviewer 28: I second @johnlanni here. I expect that a significant amount of Remote MCP Servers will provide access via APIs to their users' data.

3. Token rotation **MUST** be implemented for enhanced security.  
4. Token lifetimes **MUST** be limited based on security requirements.

## 3. Best Practices

### 3.1 Authorization servers

Almost universally, MCP server developers should prefer integrating with existing AS (as
defined in [Section 2.4](#24-authorization-server-integration)) rather than colocate or
support delegated authoriziation (as defined in
[Section 2.5](#25-delegated-authorization-or-colocated-as-flows)). This significantly
reduces the burden on developers and reduces the risk of incorrect token handling
implementations.

### 3.2 Local clients as Public OAuth 2.1 Clients

We strongly recommend that local clients implement OAuth 2.1 as a public client:

1. Utilizing code challenges (PKCE) for authorization requests to prevent interception attacks.  
2. Implementing secure token storage appropriate for the local system.  
3. Following token refresh best practices to maintain sessions.  
4. Properly handling token expiration and renewal.

### 3.3 Dynamic Client Registration

Since clients do not know the set of MCP servers in advance, we strongly recommend the implementation

> of dynamic client registration. This allows applications to automatically register with the MCP
> server, and removes the need for users to obtain client IDs manually.

Reviewer 10: ```suggestion
of dynamic client registration. This allows applications to automatically register with the authorization server referenced by the MCP server, and removes the need for users to obtain client IDs manually.
```
Technically, dynamic client registration should occur on the Authorization Server referenced by the MCP server—not on the MCP server itself. This would help readers better align their understanding with OAuth 2.1 practices.

| 403         | Forbidden    | Invalid scopes or insufficient permissions |
| 400         | Bad Request  | Malformed authorization request            |

### 2.12 Implementation Requirements

1. Implementations **MUST** follow OAuth 2.1 security best practices.  

> 2. PKCE is **REQUIRED** for all clients.  

Reviewer 23: This bullet point is already covered by point 1, given that PKCE is required for all clients using the authorization code flow in OAuth 2.1.
